### PR TITLE
WIP: Snapshot blob store improvements

### PR DESCRIPTION
**Pull request by abeyad**

### CODE REVIEW

**Comment by abeyad** [comment]
@ywelsch this doesn't compile yet, as I haven't fully gone through removing `SnapshotName`, but just to give you an idea.  Removing `SnapshotName` is only in the third commit.

**Comment by abeyad** [comment]
@ywelsch I've updated the PR with improvements.  In particular, the `Repository` APIs now mostly use `SnapshotId` and we are now using snapshot name + uuid to name the blobs in the repository.  I still need to fix some failing tests and add more tests.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 1]
```diff
@@ -104,7 +105,7 @@ protected void masterOperation(final GetSnapshotsRequest request, ClusterState s
                     }
                 }
                 for (String snapshot : snapshotsToGet) {
-                    SnapshotId snapshotId = new SnapshotId(request.repository(), snapshot);
+                    final SnapshotId snapshotId = snapshotsService.resolveSnapshotId(new SnapshotName(request.repository(), snapshot));
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: This means that we query the repository for every name in this loop? Wouldn't it make sense to add a resolveSnapshotIds method that resolves all of them in one go?

Comment (abeyad) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: Good idea, i'll add that.

Comment (abeyad) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: Although, to really make it worth it efficiency wise, the Repository interface should implement it as well, so we are only reading from disk once.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java [hunk 2]
```diff
@@ -196,12 +197,13 @@ private SnapshotsStatusResponse buildResponse(SnapshotsStatusRequest request, Li
...
-                for (String snapshotName : request.snapshots()) {
-                    SnapshotId snapshotId = new SnapshotId(request.repository(), snapshotName);
-                    if (currentSnapshotIds.contains(snapshotId)) {
+                for (String name : request.snapshots()) {
+                    SnapshotName snapshotName = new SnapshotName(request.repository(), name);
+                    if (currentSnapshotNames.contains(snapshotName)) {
                         // This is a snapshot the is currently running - skipping
                         continue;
                     }
+                    final SnapshotId snapshotId = snapshotsService.resolveSnapshotId(snapshotName);
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java]: again, this should be resolved outside of the loop. Also we have to be careful about when we are resolving the names (here we are using generic thread pool).

File: core/src/main/java/org/elasticsearch/cluster/RestoreInProgress.java [hunk 3]
```diff
@@ -77,12 +77,12 @@ public RestoreInProgress(Entry... entries) {
...
      *
-     * @param snapshotId snapshot id
+     * @param snapshotName snapshot id
      * @return restore metadata or null
      */
-    public Entry snapshot(SnapshotId snapshotId) {
+    public Entry snapshot(SnapshotName snapshotName) {
         for (Entry entry : entries) {
-            if (snapshotId.equals(entry.snapshotId())) {
+            if (snapshotName.equals(entry.snapshotId())) {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/RestoreInProgress.java]: this looks weird, comparing name and id?

File: core/src/main/java/org/elasticsearch/cluster/metadata/SnapshotName.java [hunk 4]
```diff
@@ -21,23 +21,27 @@
...
 import org.elasticsearch.common.io.stream.StreamOutput;
-import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.io.stream.Writeable;
 
 import java.io.IOException;
+import java.util.Objects;
 
 /**
- * Snapshot ID - repository name + snapshot name
+ * Snapshot Name - repository name + snapshot name
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/metadata/SnapshotName.java]: Snapshot Name - repository name + snapshot name ;-)

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 5]
```diff
@@ -489,6 +445,94 @@ private MetaData readSnapshotMetaData(SnapshotId snapshotId, Version snapshotVer
...
+    }
+
+    /**
+     * Writes snapshot index file.  This file contains the authoritative set of snapshots
+     * in the repository.
+     *
+     * @param snapshots snapshots to write to the snapshot index file
+     * @throws IOException I/O errors
+     */
+    private void writeSnapshot(Collection<SnapshotId> snapshots) throws IOException {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: why `writeSnapshot` and not `writeSnapshotList`?

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestRecoveryAction.java [hunk 6]
```diff
@@ -90,7 +90,8 @@ protected Table getTableWithHeader(RestRequest request) {
                 .addCell("source_host", "alias:shost;desc:source host")
                 .addCell("target_host", "alias:thost;desc:target host")
                 .addCell("repository", "alias:rep;desc:repository")
-                .addCell("snapshot", "alias:snap;desc:snapshot")
+                .addCell("snapshot_name", "alias:snapName;desc:snapshot_name")
+                .addCell("snapshot_uuid", "alias:snapUUID;desc:snapshot_uuid")
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestRecoveryAction.java]: we only have index name (and not index uuid) in this one, wonder if we need snapshot uuid here...

File: core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java (hunk 2 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java]: The original implementation also ends up calling `repository#readSnapshot` which goes to disk to read snapshot info, so presumably with this operation executing in the generic thread pool, it won't cause any more blocking than the current implementation?

**Pull request closed without merging**